### DISCUSSION TITLE
Is it possible to have multiple of same Identity Provider in one realm?

### DISCUSSION THREAD

**Discussion by canilsenlogiq** [keycloak/keycloak#15172] [Q&A]
Hello.

Is it possible to have multiple of same Identity Provider setup in one realm?
After adding a Microsoft IdP on one of our realms, I see that the Microsoft option have disappeared from the _"Add Provider"_ dropdown list.

If not, is there a workaround or another solution? We have multiple customers that want to our portal to be integrated with their own AD solution.

Thanks for reading.

**Comment by kkcmadhu** [comment]
you can add multiple  oauth/openid providers, not sure about micorsoft AD

**Reply by canilsenlogiq** [reply]
Thank you for your comment, I will try to add Azure AD as a OpenID Connect v1.0 IdP and see how that works out.

**Reply by canilsenlogiq** [reply]
I tried OpenID Connect v1.0 as you suggested, and I could create multiple IdP _connecting_ to different Microsoft/Azure ADs.
However, this will create multiple _"Sign in with"_ boxes/buttons on the login form. Meaning if we have lot's of customers that have their own Microsoft AD, we would end up with tons of buttons on the login screen. The goal is to have one button for Microsoft alone. 

Do you have any thought/tips?

**Comment by kkcmadhu** [comment]
i would recommend, having one realm per customer...

**Reply by canilsenlogiq** [reply]
Not sure if/how that is possible when we setup our artifacts in wildfly standalone to point towards one realm.
It is also very different to how our Keycloak, realms, attributes/roles and user groups are setup today.
```
<secure-deployment name="customer-gui.war">
  <realm>*****</realm>
  <resource>******</resource>
  <auth-server-url>https://*****</auth-server-url>
  <credential name="secret">*****</credential>
</secure-deployment>
```

**Reply by kkcmadhu** [reply]
well in that case you can see if  kc_idp_hint query param can help you.

if your customers  (who use differnt identity provider) can pass the hint they will not be seeing the normal keycloak login page and will be sent to the identity providers login page directly.. see if that can help you.

more about it here  
https://www.keycloak.org/docs/latest/server_admin/#_client_suggested_idp
https://www.keycloak.org/docs/latest/securing_apps/index.html#_params_forwarding

**Reply by Shaijil** [reply]
@canilsenlogiq 

I am in a similar situation, My goal is to allow users from different Azure AD organizations to access my application using their respective enterprise account. 
Kindly let me know if you have any update/recommendation in this regard.  
Thanks for reading.

**Reply by canilsenlogiq** [reply]
@Shaijil 
The way we ended up doing it was having only one Microsoft IdP setup towards our own Azure AD. Then we send out guest invites to external users from other ADs to join our Azure AD. It has worked great so far, but have obvious drawbacks such as:

- Need to have your own AD.
- Have to invite one and one user, so if ever a company with thousands of users needs to join, it will be tons of work.
- The customer cannot control any access except kicking the user out of their own AD(That I know of).

Personally I think setting up multiple OpenID Connect IdPs, and creating an URL for each organization that forwards/maps with the kc_idp_hint query param is also a good solution.

**Reply by Shaijil** [reply]
@canilsenlogiq 
Thanks for sharing your experience and insights. It's helpful to understand the different approaches you've explored.
I agree that inviting individual users is quite difficult, especially for large organizations. The option to invite/sync entire domains would be ideal. I figured Azure has option for bulk upload though.

**Reply by the-corner-guy** [reply]
> @Shaijil The way we ended up doing it was having only one Microsoft IdP setup towards our own Azure AD. Then we send out guest invites to external users from other ADs to join our Azure AD. It has worked great so far, but have obvious drawbacks such as:
> 
> * Need to have your own AD.
> * Have to invite one and one user, so if ever a company with thousands of users needs to join, it will be tons of work.
> * The customer cannot control any access except kicking the user out of their own AD(That I know of).
> 
> Personally I think setting up multiple OpenID Connect IdPs, and creating a URL for each organization that forwards/maps with the kc_idp_hint query param is also a good solution.


@canilsenlogiq 
How was your setup, one realm per customer, or all customers go to the same realm

I have a similar use case, but this is only possible if all users go to the same realm, and if I have to maintain multi-tenancy due to security reasons what are my options

**Reply by canilsenlogiq** [reply]
> > @Shaijil _... snip ..._
> 
> @canilsenlogiq How was your setup, one realm per customer, or all customers go to the same realm
> 
> I have a similar use case, but this is only possible if all users go to the same realm, and if I have to maintain multi-tenancy due to security reasons what are my options

We have one realm for all customers. Our realm has one `Microsoft IdP` which _pairs_ to our own Azure AD. I'm not an admin for our Azure AD, so I don't know the AD details. But the way I understand it, is that we invite other _emails_ to join our AD as guests. Then they can login with the `Or sign in with "Microsoft"` button on the same basis as our company users already in the AD.

**Comment by danielFesenmeyer** [comment]
Afaik, you can have only one IDP instance for each "Social Provider" (like Google, Microsoft etc).
But you can create multiple standard openid IDP instances. However, note that Azure AD (Entra ID) has some behavior which is not supported by the standard OpenId provider (see e.g. https://github.com/keycloak/keycloak/issues/26192)
